bolt/src/Passes/AdrPass.cpp

```diff
@@ -0,0 +1,73 @@
+//===--- AdrPass.cpp ------------------------------------------------------===//
```

Comment:
In this pass we are fixing a linker problem, right?

From what I can tell, the user used the wrong program size model and caused the compiler to emit an ADR instruction when it shouldn't, because the target of this ADR could be out-of-range. Either that or the code was written in assembly-language directly, and the programmer used the ADR instruction incorrectly.

I'm thinking that the linker, when dealing with code that contains this restriction, could violate the range as well and fail with a relocation out-of-bounds error. Do you know if the linker will do the same relaxation to compensate for the limited ADR range? I don't think so, right? I don't think the linker is allowed to expand code inside a function in that way because expanding code may change offsets and break addresses that are already resolved by the assembler.

So my question would be, what's preventing this input from crashing the linker, instead of crashing BOLT? If it *could* crash the linker, then we shouldn't fix in BOLT, because it is only a matter of time when the linker is going to start crashing (and it is probably not crashing by luck). If it can't crash the linker, then I'm curious to understand why, and why is BOLT violating some assumption that the linker is not.

I would also rename this pass to ADRRelaxationPass, so it is more clear its intent.

Reply:
This is very good question, which I didn't exam to be fare. I'm not 100% sure why it happens, it seems to be ld problem for real. But the problem is quite strange. For the quite big binary (redis server) with stack canary enabled I've got only 2 functions which are get canary's value with ADR instruction, not ADRP. But the thing is the ADR instruction is maintained page aligned, so ld still generated smth like this:
adr x0, #ff0000 //Page aligned
ldr x0, [x0, #page_offset]

Reply:
@rafaelauler The errata 843419 file:///tmp/mozilla_yota90/Cortex-A53_MPCore_Software_Developers_Errata_Notice_21.pdf might cause such a behavior.. It's 3.30am, I will probably check it tomorrow..

Reply:
@rafaelauler UPD I've checked the binary, in both cases the instruction address is ended with ff8 and ffc, this is exactly what errata says. The LD patches it, so BOLT is not compatible with -mfix-cortex-a53-843419 option. Of course the addresses are completely changed by the bolt and we could replace them by the ADRP + ADD change (although for this errata the add would always be 0, since LD just changed ADRP to ADR which stays page-aligned). I should probalby reformat the patch to check that the the extracted value is page-aligned and just replace ADR with ADRP (and target expression) at least for now. 
BTW this is the second time I'm thinking about such errata cases, I assume the LLVM won't fix them after BOLT? Especially if we have more compex case where ADRP could not be changed to ADR due to long distance. The other problem I thought was Intel "JCC Erratum", where extra nops could be needed to avoid it.. As I understand (however I'm not sure) currently such cases are ignored by bolt and llvm? If so are there any plans to support such problems? Thank you!

Reply:
Oh, now it's clear what's happening, thanks for explaining the context! Can you update the pass description to explain that we are fixing LD erratum? If I got it correctly, LD is patching an instruction to compensate for erratum 843419 and, in doing so, it is reducing the range, which is fine because LD knows the target is within range. But then BOLT comes and re-does the layout, and now we need to undo the changes done by LD and possibly recalculate which places need to be fixed according to the errata. But at the very minimum this pass is required, so the binary doesn't crash even in processors not affected by the errata.

In general, errata corrections that requires knowledge of the exact offset where an instruction is are a bit awkward to implement. The compiler usually can't do that, but the assembler can. Essentially, this requires fiddling with the assembler (LLVM's integrated assembler, the MCStreamer class), because that is one of the few components in the whole pipeline that has precise layout knowledge. When you use jcc erratum's in clang, it is essentially passing a flag to LLVM's assembler to do all the work.

In BOLT, we use LLVM's assembler as well, but whenever we need to implement an erratum fix, that means changes to it, which are outside of our control. https://reviews.llvm.org/D97982 is an example of one special assembler fragment that we introduced to aid in implementing these tricky layout fixes, but it was not accepted upstream.

In a related problem, we need precise layout to compute stubs (LongJmp.cpp). In this case, we do it outside of LLVM's assembler (in our own pass). This is quite expensive because we need to perfectly mimic the layout of the assembler and then perform adjustments. It works for fixing out-of-range control flow instructions, but it wouldn't work for erratum 843419 because we are not computing exact addresses of each instruction. Our granularity for that pass is basic-block level, so we know where each basic block is, but we don't know where each instruction is, which would be very expensive and is a task that is indeed better suited for the assembler.

If  you don't have this processor model in the fleet, disable the erratum fix in ld. If it is in the fleet and it is crashing regularly and a fix is required, try to look for how to do these NOP paddings in MCStreamer (for example, here https://github.com/facebookincubator/BOLT/blob/main/bolt/src/BinaryEmitter.cpp#L447 is an area where you might interact with the assembler (Streamer object) by requesting it to conditionally pad the instruction with NOPs if certain layout conditions are met (depending on the address, which is something the assembler can do for you).